### PR TITLE
Trimming: Bind memory2

### PR DESCRIPTION
**Pull request by locke-lunarg**
https://github.com/LunarG/gfxreconstruct/issues/273

- vkBindImageMemory2 and vkBindBufferMemory2
- VkBindImageMemorySwapchainInfoKHR

Test code: https://github.com/locke-lunarg/Vulkan-Tools/tree/locke-bind_memory2

### CODE REVIEW

**Review by patrick-lunarg — CHANGES_REQUESTED** [review]

File: framework/generated/generated_encode_pnext_struct.cpp [hunk 1]
```diff
@@ -1021,5 +1021,2235 @@ void EncodePNextStruct(ParameterEncoder* encoder, const void* value)
     }
 }
 
+std::vector<std::vector<uint8_t>> CopyPNextStructs(const void* value)
```
[the remark this answers is not in the document]
Comment (locke-lunarg) [reply framework/generated/generated_encode_pnext_struct.cpp]: @patrick-lunarg  Thanks. I changed to use `HandleUnwrapMemory` and `CopyPNextStruct` to track pNext.

File: framework/encode/vulkan_state_writer.cpp [hunk 2]
```diff
@@ -1572,16 +1596,39 @@ void VulkanStateWriter::WriteImageMemoryState(const VulkanStateTable& state_tabl
             parameter_stream_.Reset();
 
             // Write memory bind command.
-            encoder_.EncodeHandleIdValue(device_wrapper->handle_id);
-            encoder_.EncodeHandleIdValue(wrapper->handle_id);
-            encoder_.EncodeHandleIdValue(memory_wrapper->handle_id);
-            encoder_.EncodeVkDeviceSizeValue(wrapper->bind_offset);
-            encoder_.EncodeEnumValue(VK_SUCCESS);
+            if (!wrapper->bind_pnext)
```
Comment (bradgrantham-lunarg) [inline framework/encode/vulkan_state_writer.cpp:1599]: Please make this wrapper->bind_pnext == nullptr to be consistent with other use

Comment (locke-lunarg) [reply framework/encode/vulkan_state_writer.cpp]: fixed

File: framework/encode/custom_vulkan_struct_handle_wrappers.cpp [hunk 3]
```diff
@@ -124,5 +124,32 @@ void UnwrapStructHandles(VkWriteDescriptorSet* value, HandleUnwrapMemory* unwrap
...
 }
 
+const void* TrackPNextStruct(const void* value, HandleUnwrapMemory* unwrap_memory)
+{
+    const void*              ret        = nullptr;
+    const VkBaseInStructure* pnext      = nullptr;
+    const VkBaseInStructure* last_pnext = nullptr;
+    auto                     base       = reinterpret_cast<const VkBaseInStructure*>(value);
+    bool                     first      = true;
+    while (base)
```
[the remark this answers is not in the document]
Comment (locke-lunarg) [reply framework/encode/custom_vulkan_struct_handle_wrappers.cpp]: fixed

File: framework/encode/vulkan_state_tracker.cpp [hunk 4]
```diff
@@ -208,30 +209,41 @@ void VulkanStateTracker::TrackBufferDeviceAddress(VkDevice device, VkBuffer buff
...
     assert((device != VK_NULL_HANDLE) && (buffer != VK_NULL_HANDLE) && (memory != VK_NULL_HANDLE));
 
     auto wrapper            = reinterpret_cast<BufferWrapper*>(buffer);
     wrapper->bind_device    = reinterpret_cast<DeviceWrapper*>(device);
     wrapper->bind_memory_id = GetWrappedId(memory);
     wrapper->bind_offset    = memoryOffset;
+    wrapper->bind_pnext     = nullptr;
+    wrapper->bind_pnext_memory.Reset();
+
+    if (pnext)
```
[the remark this answers is not in the document]
Comment (locke-lunarg) [reply framework/encode/vulkan_state_tracker.cpp]: fixed

File: framework/encode/vulkan_state_tracker.cpp [hunk 5]
```diff
@@ -208,30 +209,41 @@ void VulkanStateTracker::TrackBufferDeviceAddress(VkDevice device, VkBuffer buff
...
+    assert((device != VK_NULL_HANDLE) && (image != VK_NULL_HANDLE));
 
     auto wrapper            = reinterpret_cast<ImageWrapper*>(image);
     wrapper->bind_device    = reinterpret_cast<DeviceWrapper*>(device);
     wrapper->bind_memory_id = GetWrappedId(memory);
     wrapper->bind_offset    = memoryOffset;
+    wrapper->bind_pnext     = nullptr;
+    wrapper->bind_pnext_memory.Reset();
+
+    if (pnext)
```
[the remark this answers is not in the document]
Comment (locke-lunarg) [reply framework/encode/vulkan_state_tracker.cpp]: fixed

File: framework/encode/vulkan_state_writer.cpp [hunk 6]
```diff
@@ -1200,7 +1200,8 @@ void VulkanStateWriter::ProcessImageMemory(const DeviceWrapper*
         const DeviceMemoryWrapper* memory_wrapper = snapshot_entry.memory_wrapper;
         const uint8_t*             bytes          = nullptr;
 
-        assert((image_wrapper != nullptr) && (memory_wrapper != nullptr));
+        assert((image_wrapper != nullptr) && ((image_wrapper->is_swapchain == true && memory_wrapper == nullptr) ||
```
[the remark this answers is not in the document]
Comment (locke-lunarg) [reply framework/encode/vulkan_state_writer.cpp:1203]: fixed

File: framework/encode/vulkan_handle_wrappers.h [hunk 7]
```diff
@@ -174,7 +175,10 @@ struct DeviceMemoryWrapper : public HandleWrapper<VkDeviceMemory>
 
 struct BufferWrapper : public HandleWrapper<VkBuffer>
 {
-    DeviceWrapper*   bind_device{ nullptr };
+    DeviceWrapper*     bind_device{ nullptr };
+    const void*        bind_pnext;
+    HandleUnwrapMemory bind_pnext_memory; // Gobal HandleUnwrapMemory could be reset anytime, so it should have its own
```
[the remark this answers is not in the document]
Comment (locke-lunarg) [reply framework/encode/vulkan_handle_wrappers.h]: fixed

File: framework/encode/custom_vulkan_struct_handle_wrappers.cpp [hunk 8]
```diff
@@ -124,5 +124,32 @@ void UnwrapStructHandles(VkWriteDescriptorSet* value, HandleUnwrapMemory* unwrap
     }
 }
 
+const void* TrackPNextStruct(const void* value, HandleUnwrapMemory* unwrap_memory)
+{
+    const void*              ret        = nullptr;
+    const VkBaseInStructure* pnext      = nullptr;
```
[the remark this answers is not in the document]
Comment (locke-lunarg) [reply framework/encode/custom_vulkan_struct_handle_wrappers.cpp]: fixed

File: framework/encode/custom_vulkan_struct_handle_wrappers.cpp [hunk 9]
```diff
@@ -124,5 +124,31 @@ void UnwrapStructHandles(VkWriteDescriptorSet* value, HandleUnwrapMemory* unwrap
...
+    const void*        ret        = nullptr;
+    VkBaseInStructure* pnext      = nullptr;
+    VkBaseInStructure* last_pnext = nullptr;
+    auto               base       = reinterpret_cast<const VkBaseInStructure*>(value);
+    bool               first      = true;
+    while (base != nullptr)
+    {
+        // Use CopyPNextStruct, instead of UnwrapPNextStructHandles because it doesn't need to wrap handles.
+        pnext = CopyPNextStruct(base, unwrap_memory);
+        if (first == true)
```
Comment (bradgrantham-lunarg) [inline framework/encode/custom_vulkan_struct_handle_wrappers.cpp]: `if (first)`

File: framework/encode/vulkan_state_writer.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (bradgrantham-lunarg) [reply framework/encode/vulkan_state_writer.cpp:1203]: I mean like
```
assert((image_wrapper != nullptr) && ((image_wrapper->is_swapchain && memory_wrapper == nullptr) ||
        (!image_wrapper->is_swapchain && memory_wrapper != nullptr)));
```

File: framework/encode/vulkan_state_writer.cpp [hunk 10]
```diff
@@ -1368,7 +1369,7 @@ void VulkanStateWriter::ProcessImageMemory(const DeviceWrapper*
                 }
             }
         }
-        else
+        else if (image_wrapper->is_swapchain == false)
```
Comment (bradgrantham-lunarg) [inline framework/encode/vulkan_state_writer.cpp:1372]: `if (!image_wrapper->is_swapchain)`

File: framework/encode/vulkan_state_writer.cpp [hunk 11]
```diff
@@ -1401,73 +1402,76 @@ void VulkanStateWriter::ProcessImageMemory(const DeviceWrapper*
...
-        upload_cmd.meta_header.meta_data_id =
-            format::MakeMetaDataId(format::ApiFamilyId::ApiFamily_Vulkan, format::MetaDataType::kInitImageCommand);
-        upload_cmd.thread_id = thread_id_;
-        upload_cmd.device_id = device_wrapper->handle_id;
-        upload_cmd.image_id  = image_wrapper->handle_id;
-        upload_cmd.aspect    = snapshot_entry.aspect;
-        upload_cmd.layout    = image_wrapper->current_layout;
-
-        if (bytes != nullptr)
+        if (image_wrapper->is_swapchain == false)
```
Comment (bradgrantham-lunarg) [inline framework/encode/vulkan_state_writer.cpp:1405]: `if (!image_wrapper->is_swapchain)`

File: framework/encode/vulkan_state_writer.cpp [hunk 12]
```diff
@@ -1552,7 +1575,8 @@ void VulkanStateWriter::WriteImageMemoryState(const VulkanStateTable& state_tabl
         // Perform memory binding.
         const DeviceMemoryWrapper* memory_wrapper = state_table.GetDeviceMemoryWrapper(wrapper->bind_memory_id);
 
-        if (memory_wrapper != nullptr)
+        if ((wrapper->is_swapchain == true && memory_wrapper == nullptr && wrapper->bind_device != nullptr) ||
+            (wrapper->is_swapchain == false && memory_wrapper != nullptr))
```
Comment (bradgrantham-lunarg) [inline framework/encode/vulkan_state_writer.cpp:1579]:
```
if ((wrapper->is_swapchain && memory_wrapper == nullptr && wrapper->bind_device != nullptr) ||
            (!wrapper->is_swapchain && memory_wrapper != nullptr))
```

**Review by bradgrantham-lunarg — CHANGES_REQUESTED** [review]

File: framework/encode/custom_vulkan_struct_handle_wrappers.cpp (hunk 9 above)
Comment (locke-lunarg) [reply framework/encode/custom_vulkan_struct_handle_wrappers.cpp]: Fixed. Removed `first` and used `if (ret == nullptr)`

File: framework/encode/vulkan_state_writer.cpp (hunk 10 above)
Comment (locke-lunarg) [reply framework/encode/vulkan_state_writer.cpp]: Fixed. Also changed `is_swapchain` to `is_swapchain_image`

File: framework/encode/vulkan_state_writer.cpp (hunk 11 above)
Comment (locke-lunarg) [reply framework/encode/vulkan_state_writer.cpp]: Fixed. Also changed `is_swapchain` to `is_swapchain_image`

File: framework/encode/vulkan_state_writer.cpp (hunk 12 above)
Comment (locke-lunarg) [reply framework/encode/vulkan_state_writer.cpp]: Fixed. Also changed `is_swapchain` to `is_swapchain_image`

**Review by bradgrantham-lunarg — APPROVED** [review]

**Review by patrick-lunarg — APPROVED** [review]

**Pull request merged**